harness/tests/determined/pytorch/experimental/test_torch_batch_process.py

```diff
@@ -0,0 +1,627 @@
+    "data_length, batch_size, checkpoint_interval, rank, num_slots, "
+    "steps_completed, expected_process_batch_call_count",
+    [  # total batches for worker (5), completed (2), iterate for 3 batches
+        [50, 10, 1, 0, 1, 2, 3],
+        # total batches for worker 0 (3), completed (2), iterate for 1 batches
+        [50, 10, 1, 0, 2, 2, 1],
+        # total batches for worker 1 (2), completed (2), iterate for 0 batches
+        [50, 10, 1, 1, 2, 2, 0],
+    ],
+    ids=[
+        "Single-slot; total_batches=5; completed_batches=2",
+        "Multi-slot; chief; total_batches=5; completed_batches=2",
+        "Multi-slot; worker; total_batches=5; completed_batches=2",
+    ],
+)
+def test_torch_batch_process_skip_completed_batches(
+    mock_synchronize_and_checkpoint: unittest.mock.MagicMock,
+    mock_initialize_default_inference_context: unittest.mock.MagicMock,
+    mock_load_state: unittest.mock.MagicMock,
+    data_length: int,
+    batch_size: int,
+    checkpoint_interval: int,
+    rank: int,
+    num_slots: int,
+    steps_completed: int,
+    expected_process_batch_call_count: int,
+) -> None:
+    with test_util.set_mock_cluster_info(
+        DEFAULT_ADDRS, rank, num_slots, latest_checkpoint="fake_latest_checkpoint"
+    ):
+        # Mock core context as _dummy_init would call storage functions
+        mock_initialize_default_inference_context.return_value = _get_core_context(rank=rank)
+        index_dataset = IndexData(data_length)
+
+        MyProcessorCLS = unittest.mock.Mock()
+        my_processor_instance = unittest.mock.Mock()
+        MyProcessorCLS.return_value = my_processor_instance
+
+        mock_load_state.return_value = {
+            "steps_completed": steps_completed,
+            "default_output_uuid": "abc",
+        }
+
+        experimental.torch_batch_process(
+            dataset=index_dataset,
+            batch_processor_cls=MyProcessorCLS,
+            batch_size=batch_size,
+            checkpoint_interval=checkpoint_interval,
+        )
+
+        assert my_processor_instance.process_batch.call_count == expected_process_batch_call_count
+        # on_finish should only be called once
+        assert my_processor_instance.on_finish.call_count == 1
+
+
+@unittest.mock.patch(
+    "determined.pytorch.experimental._torch_batch_process._initialize_default_inference_context"
+)
+@unittest.mock.patch(
+    "determined.pytorch.experimental._torch_batch_process._synchronize_and_checkpoint"
+)
+@pytest.mark.parametrize(
+    "data_length, batch_size, checkpoint_interval, rank, num_slots, "
+    "max_batches, expected_process_batch_call_count",
+    [  # max_batches (2) < total batches (3) -> iterate for 2 batches
+        [50, 10, 1, 0, 2, 2, 2],
+        # max_batches (10) > total batches (3) -> iterate for 3 batches
+        [50, 10, 1, 0, 2, 10, 3],
+    ],
+    ids=["Valid max_batches arg", "Invalid max_batches arg"],
+)
+def test_torch_batch_process_only_use_valid_max_batches_arg(
```

Comment:
I don't fully understand what this tests. Don't we know how 

```python
if max_batches > expected_process_batch_call_count:
```

is going to turn out?

Reply:
Updated to break tests into one for valid max_batch arg and an invalid max_batch arg. Thank you!